sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,247 @@
+	#objects: AsyncCache<ObjectCacheEntry>;
+	#functions: AsyncCache<MoveFunctionEntry>;
+
+	constructor({ createCache = InMemoryCache.createCache }: ObjectCacheOptions = {}) {
+		this.#createCache = createCache;
+		this.#objects = this.#createCache('objects');
+		this.#functions = this.#createCache('functions');
+	}
+
+	getObjects: NonNullable<TransactionBlockDataResolverPlugin['getObjects']> = async (ids, next) => {
+		const results = new Map<string, ObjectCacheEntry>();
+
+		await Promise.all(
+			ids.map(async (id) => {
+				const cachedObject = await this.#objects.get(id);
+
+				if (cachedObject) {
+					results.set(id, cachedObject);
+				}
+			}),
+		);
+		const missingIds = ids.filter((id) => !results.has(id));
+
+		if (missingIds.length > 0) {
+			const newObjects = await next(missingIds);
+
+			await Promise.all(
+				newObjects.map(async (newObject) => {
+					await this.addObject(newObject);
+					results.set(newObject.objectId, newObject);
+				}),
+			);
+		}
+
+		return ids.map((id) => results.get(id)!);
+	};
+
+	getMoveFunctionDefinition: NonNullable<
+		TransactionBlockDataResolverPlugin['getMoveFunctionDefinition']
+	> = async (ref, next) => {
+		const functionName = `${normalizeSuiAddress(ref.package)}::${ref.module}::${ref.function}`;
+		const cached = await this.#functions.get(functionName);
+		if (cached) {
+			return cached;
+		}
+
+		const functionDefinition = await next(ref);
+
+		return await this.addFunction(functionDefinition);
+	};
+
+	async clearCache() {
+		await Promise.all([this.#objects.clear(), this.#functions.clear()]);
+	}
+
+	async addObject(object: ObjectCacheEntry) {
+		await this.#objects.set(object.objectId, object);
+		return object;
+	}
+
+	async addFunction(functionEntry: MoveFunctionEntry) {
+		const pkg = normalizeSuiAddress(functionEntry.package);
+		const functionName = `${pkg}::${functionEntry.module}::${functionEntry.function}`;
+		const entry = {
+			...functionEntry,
+			package: pkg,
+		};
+
+		await this.#functions.set(functionName, entry);
+
+		return entry;
+	}
+
+	async invalidateObject(id: string) {
+		await this.#objects.delete(id);
+	}
+
+	async applyEffects(effects: typeof bcs.TransactionEffects.$inferType) {
+		if (!effects.V2) {
+			throw new Error(`Unsupported transaction effects version ${effects.$kind}`);
+		}
+
+		const { lamportVersion, changedObjects } = effects.V2;
+
+		await Promise.all(
+			changedObjects.map(async ([id, change]) => {
+				if (change.outputState.NotExist) {
+					await this.invalidateObject(id);
+				} else if (change.outputState.ObjectWrite) {
+					const [digest, owner] = change.outputState.ObjectWrite;
+
+					await this.addObject({
+						objectId: id,
+						digest,
+						version: lamportVersion,
+						owner: owner.AddressOwner ?? owner.ObjectOwner ?? null,
+						initialSharedVersion: owner.Shared?.initialSharedVersion ?? null,
+					});
+				}
+			}),
+		);
+	}
+}
+
+export class CachingTransactionBlockExecutor extends ObjectCache {
```

Comment:
I wouldn't make this extend the class just to keep the intelligence methods minimal, but don't feel strongly